Automattic/wp-calypso
# Adds a plugn to extract webpack's manifest into a global

### Background

Webpack has the concept of [The Manifest](https://webpack.js.org/concepts/manifest/), but it is actually a combination of webpack's runtime and the actual manifest (as in, the list of chunks). Usually this manifest is embedded in the main HTML file, meaning that every single user (even if they have all assets cached) have to download and execute it before it can request the first chunk.

`mini-css-extract-plugin` makes this way worse because it adds another two copies of the list of chunks to the embedded manifest (see #42979)

### Changes

This *experimental* plugin splits the manifest into runtime and the actual manifest (i.e the list of chunks only). This way we can embed the list of chunks (the only part that usually changes between compilations) and extract the runtime to an external file that the browser can cache. This saves some bytes in the critical path.

By extracting the manifest to a global variable it allows `mini-css-extract-plugin` to reuse it, so it doesn't embed the manifest twice again.

Both effects combined should save ~50kb (before gzip) from the critical path for every request.

### How it works

Webpack's manifest has a function like

```
function jsonpScriptSrc(chunkId) {
    return __webpack_require__.p + ({1: '1.js', 2:'2.js'}[chunkId]) +".min.js"
}
```

that transforms a chunkId (eg: `1`) into a file (eg: `1.js`). The map chunkId->file is the manifest, and it can be quite big (>500 entries). This map is composed by webpack by combining the output filename (something like `[name].[chunkhash].min.js`) with the actual list of chunks generated.

This plugin works by hooking into the main template rendering to intercept the variable interpolation in the asset name (eg: `[name].js`). For each var, it will generate a call to a global function (eg: `window.__WEBPACK_MANIFEST.get("name")(chunkId)+".js").

Also, for each var, it will call Webpack variable interpolation logic with just that variable (`getAssetsPath('[name]')), capture the output and save it in a global Map (`window.__WEBPACK_MANIFEST.set('name', function(chunkId){...})`) so it returns the correct variable value for a given `chunkdId`.

Then it will add a second file to the `runtime` chunk with all `window.__WEBPACK_MANIFEST.set()` calls as content. This will allow future improvements were we can inline the manifest (the list of `window.__WEBPACK_MANIFEST.set()` calls) and move the runtime to an external, cacheable script.

So with the above example, we'll end up with:

```
//runtime.js
function jsonpScriptSrc(chunkId) {
    return __webpack_require__.p + window.__WEBPACK_MANIFEST.get("name")(chunkId) +".min.js"
}

//manifest.js
window.__WEBPACK_MANIFEST=new Map();
window.__WEBPACK_MANIFEST.set('name', function(chunkId) {return ({1: '1.js', 2:'2.js'}[chunkId])});
```

The main benefit is that any other plugin (eg: `mini-css-extract-plugin`) needs a way to transform a filename template like `[name].css` it won't need another copy of the full chunkdId->name map. It will reuse the `window.__WEBPACK_MANIFEST.get("name")` call, resulting in a much smaller runtime.

You can see all this in action by loading the live branch and looking for an inline script with the content `__WEBPACK_MANIFEST`. 
 
### Notes

This plugin is right at the limit of my webpack knowledge. I'd like to have some webpack-expert eyes to make sure I didn't misunderstand webpack internals.

### Testing instructions

* Check out the branch and verify `yarn start` still works.
* Load live branch and click around between sections to make sure everything loads properly (including CSS).
* Test it on IE as well.

Fixes #42979
